src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -42,11 +45,144 @@ public void RegisterReceiver(Receiver r)
 
         private static readonly Telemetry Singleton = new Telemetry();
 
-        private Telemetry(){}
+        internal Telemetry(){}  // This is an internal constructor to build isolated unit test instance
 
         public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private ConcurrentDictionary<Tuple<string, string>, EventBase> EventsInProgress = new ConcurrentDictionary<Tuple<string, string>, EventBase>();
+
+        private ConcurrentDictionary<string, List<EventBase>> CompletedEvents = new ConcurrentDictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            if (requestId != null)
```

Comment:
The way I see it, its a perf issue, you are doing a whole bunch of work in start and stop event which is not required if there is no receiver. save memory and processing cycles.